src/core/geometry/qgscompoundcurve.cpp

```diff
@@ -1414,3 +1414,94 @@ void QgsCompoundCurve::swapXy()
   }
   clearCache();
 }
+
+double QgsCompoundCurve::distanceBetweenVertices( QgsVertexId fromVertex, QgsVertexId toVertex ) const
+{
+  // Convert QgsVertexId to simple vertex numbers for compound curves (single ring, single part)
+  if ( fromVertex.part != 0 || fromVertex.ring != 0 || toVertex.part != 0 || toVertex.ring != 0 )
+    return -1.0;
+
+  int fromVertexNumber = fromVertex.vertex;
+  int toVertexNumber = toVertex.vertex;
+
+  int totalVertices = numPoints();
+  if ( fromVertexNumber < 0 || fromVertexNumber >= totalVertices || toVertexNumber < 0 || toVertexNumber >= totalVertices )
+    return -1.0;
+
+  if ( fromVertexNumber == toVertexNumber )
+    return 0.0;
+
+  // Ensure fromVertex < toVertex for simplicity
+  if ( fromVertexNumber > toVertexNumber )
+  {
+    return distanceBetweenVertices( QgsVertexId( 0, 0, toVertexNumber ), QgsVertexId( 0, 0, fromVertexNumber ) );
+  }
+
+  double totalDistance = 0.0;
+
+  // Find which curves contain our vertices and accumulate distances
+  int currentVertexId = 0;
+  int fromCurve = -1, toCurve = -1;
+  int fromCurveVertex = -1, toCurveVertex = -1;
+
+  // First pass: find which curves contain from and to vertices
+  for ( int j = 0; j < mCurves.size(); ++j )
+  {
+    int nCurvePoints = mCurves.at( j )->numPoints();
+
+    // Check if fromVertex is in this curve
+    if ( fromCurve == -1 && fromVertexNumber >= currentVertexId && fromVertexNumber < currentVertexId + nCurvePoints )
+    {
+      fromCurve = j;
+      fromCurveVertex = fromVertexNumber - currentVertexId;
+    }
+
+    // Check if toVertex is in this curve
+    if ( toCurve == -1 && toVertexNumber >= currentVertexId && toVertexNumber < currentVertexId + nCurvePoints )
+    {
+      toCurve = j;
+      toCurveVertex = toVertexNumber - currentVertexId;
+    }
```

Comment:
We can `break` here for efficiency, can't we?

Reply:
nice catch! thanks